### PR TITLE
[spec] v2.3 - Optional Transaction In Action Chaining and External Link Support

### PR DESCRIPTION
**Pull request by thearyanag**
# TLDR

Currently, Actions can be chained together in successive series, but each action in the chain requires the user to sign a transaction thus creating a bad UX. 

1. A Post Request Type ( Optional Transaction ) has been added to avoid signing the transaction on each action.
2. A External-Link type has been added, to direct the user to an external website.

This would enable developers to 
 - providing better UX by avoiding individual pop-ups for each transaction, with the help of optional transactions.
 - support cases such as adding a `Learn More` or `Manage your Squads` kind of thing, with the help of an external link.
 - Later can be extended to support use cases like sign-message where each flow on the blink client end will be slightly different.

## Rationale

Blinks and actions currently require individual transactions for each action in the chain. If the user wants to move to the next action in the chain, they can't without signing a transaction. Having optional transactions would enable a better UX and open more design space for the developers. 

sRFCs :
1. https://forum.solana.com/t/srfc-32-optional-transactions-in-action-chaining/1971/18
4. https://forum.solana.com/t/blinks-cta-external-linking/2018/10

### CODE REVIEW

File: packages/actions-spec/index.d.ts [hunk 1]
```diff
@@ -75,10 +75,17 @@ export interface Action<T extends ActionType = "action"> {
   error?: ActionError;
 }
 
+/**
+ * Type of action to determine client side handling
+ */
+export type LinkedActionType = "tx" | "post";
```
[the remark this answers is not in the document]
Comment (nickfrosty) [reply packages/actions-spec/index.d.ts:81]: should we also include `external-link` in this same PR to cover this SRFC: https://forum.solana.com/t/blinks-cta-external-linking/2018/10 ?

File: packages/actions-spec/index.d.ts [hunk 2]
```diff
@@ -175,23 +182,37 @@ export interface ActionPostRequest<T = string> {
...
+  message?: string;
+  links?: {
+    next: NextActionLink;
+  };
+}
+
+/**
+ * Response body payload returned from the Action POST Request
+ */
+export type ActionPostResponse = TxResponse | PostResponse;
```
Comment (nickfrosty) [inline packages/actions-spec/index.d.ts:214]: we can clean up the types here some and make them reusable by creating an generic interface for the different POST responses, I called it `ActionResponse`:

```suggestion
/** Type of action to determine client side handling */
export type PostActionType = LinkedActionType;

/**
 * Generic response from an Action API request
 */
export interface ActionResponse {
  type?: PostActionType;
  message?: string;
  links?: {
    next: NextActionLink;
  };
}

/**
 * Response body payload returned from the Action POST Request if the action is a transaction
 */
export interface TransactionResponse extends ActionResponse {
  type?: Extract<PostActionType, "transaction">;
  transaction: string;
}

/**
 * Response body payload returned from the Action POST Request if the action is a POST request
 */
export interface PostResponse extends ActionResponse {
  type: Extract<PostActionType, "post">;
}

/**
 * Response body payload returned from the Action POST Request
 */
export type ActionPostResponse = TransactionResponse | PostResponse;
```

Comment (nickfrosty) [reply packages/actions-spec/index.d.ts:214]: this suggested code block also avoids using the short name of `Tx` and uses `Transaction` instead

**Comment by nickfrosty** [comment]
@thearyanag also, the `NextActionPostRequest` should be updated to make the `signature` optional too since with this PR, not all actions will be able to provide a signature

File: packages/actions-spec/index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (thearyanag) [reply packages/actions-spec/index.d.ts]: I think we can also include `external-link` in the same PR considering it as a low-hanging fruit.

File: packages/actions-spec/index.d.ts (hunk 2 above)
Comment (thearyanag) [reply packages/actions-spec/index.d.ts:214]: This looks much better.

**Comment by nickfrosty** [comment]
cc @tsmbl @fsher

**Comment by tsmbl** [comment]
this is good improvement, let's get it shipped

**Comment by 0xlarry** [comment]
this is great

**Comment by thearyanag** [comment]
hey @tsmbl , agreed on `contextual data` across requests. added as a `type`

**Review by nickfrosty — APPROVED** [review]
